File: projects/app/api/schema/project.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.core.project import Patch as CorePatch
from app.core.project import Project as CoreProject


class Door(BaseModel):
    id: str
    wall_id: str
    x: int = Field(ge=0)
    w: int = Field(ge=0)
    h: int = Field(ge=0)


class DoorPatch(BaseModel):
    id: str | None = None
    wall_id: str | None = None
    x: int | None = Field(None, ge=0)
    w: int | None = Field(None, ge=0)
    h: int | None = Field(None, ge=0)


class Window(BaseModel):
    id: str
    wall_id: str
    x: int = Field(ge=0)
    y: int = Field(ge=0)
    w: int = Field(ge=0)
    h: int = Field(ge=0)


class WindowPatch(BaseModel):
    id: str | None = None
    wall_id: str | None = None
    x: int | None = Field(None, ge=0)
    y: int | None = Field(None, ge=0)
    w: int | None = Field(None, ge=0)
    h: int | None = Field(None, ge=0)


class Wall(BaseModel):
    id: str
    x1: int
    y1: int
    x2: int
    y2: int

    def length(self):
        return abs(self.x1 - self.x2) + abs(self.y1 - self.y2)


class WallPatch(BaseModel):
    id: str | None = None
    x1: int | None = None
    y1: int | None = None
    x2: int | None = None
    y2: int | None = None


class Point(BaseModel):
    x: int
    y: int


class WetArea(BaseModel):
    id: str
    points: list[Point]


class WetAreaPatch(BaseModel):
    id: str | None = None
    points: list[Point] | None = None


class Content(BaseModel):
    walls: dict[str, Wall] = {}
    doors: dict[str, Door] = {}
    windows: dict[str, Window] = {}
    wet_areas: dict[str, WetArea] = {}


class ContentPatch(BaseModel):
    walls: dict[str, WallPatch | None] = {}
    doors: dict[str, DoorPatch | None] = {}
    windows: dict[str, WindowPatch | None] = {}
    wet_areas: dict[str, WetAreaPatch | None] = {}


class Patch(BaseModel):
    name: str | None = Field(None, max_length=256)
    description: str | None = Field(None, max_length=2048)
    content: ContentPatch | None = None

    def to_core(self):
        return CorePatch(**self.model_dump())


class Project(BaseModel):
    id: UUID
    account_id: str
    name: str = Field(max_length=256)
    description: str = Field(max_length=2048)
    content: Content | None = None
    published: bool = False
    created_at: datetime
    updated_at: datetime

    @staticmethod
    def from_core(project: CoreProject) -> "Project":
        return Project(**project.model_dump())

File: frontend/src/features/project/project.ts
export interface ProjectEntity {
  id: string;
}

export interface WetArea extends ProjectEntity {
  x: number;
  y: number;
  w: number;
  h: number;
}

export interface Door extends ProjectEntity {
  wallId: string;
  x: number;
  w: number;
  h: number;
}

export interface Window extends ProjectEntity {
  wallId: string;
  x: number;
  y: number;
  w: number;
  h: number;
}

export interface Wall extends ProjectEntity {
  x1: number;
  y1: number;
  x2: number;
  y2: number;
}

export interface ProjectPatch {
  name?: string;
  description?: string;
  content?: {
    walls?: Record<string, Partial<Wall> | null>;
    windows?: Record<string, Partial<Window> | null>;
    doors?: Record<string, Partial<Door> | null>;
    wetAreas?: Record<string, Partial<WetArea> | null>;
  }
}

export interface ProjectContent {
  walls: Record<string, Wall>;
  windows: Record<string, Window>;
  doors: Record<string, Door>;
  wetAreas: Record<string, WetArea>;
}

export interface Project {
  id: string;
  accountId: string;
  name: string;
  description: string;
  published: boolean;
  content: ProjectContent;
  dtCreated: string;
  dtUpdated: string;
  etag: string;
}

function applyJsonMergePatch<T>(target: T, patch: any): T {
  if (patch === null || typeof patch !== 'object' || Array.isArray(patch)) {
    return patch as T;
  }
  const result: any = target !== null && typeof target === 'object' && !Array.isArray(target)
    ? { ...target }
    : {};
  for (const key of Object.keys(patch)) {
    const patchValue = patch[key];
    if (patchValue === null) {
      delete result[key];
    }
    else if (typeof patchValue === 'object' && !Array.isArray(patchValue)) {
      const currentValue = result[key];
      result[key] = applyJsonMergePatch(currentValue, patchValue);
    }
    else {
      result[key] = patchValue;
    }
  }
  return result as T;
}

function mergePatchObjects(target: any, source: any): any {
  const result = { ...target };
  for (const key of Object.keys(source)) {
    const sourceValue = source[key];
    if (sourceValue === null) {
      result[key] = null;
    }
    else if (
      typeof sourceValue === 'object' &&
      !Array.isArray(sourceValue) &&
      typeof result[key] === 'object' &&
      !Array.isArray(result[key]) &&
      result[key] !== null
    ) {
      result[key] = mergePatchObjects(result[key], sourceValue);
    }
    else {
      result[key] = sourceValue;
    }
  }
  return result;
}


export const ProjectUtils = {
  isEmpty(project: Project): boolean {
    return (
      Object.keys(project.content.doors).length === 0 &&
      Object.keys(project.content.walls).length === 0 &&
      Object.keys(project.content.windows).length === 0 &&
      Object.keys(project.content.wetAreas).length === 0
    );
  },
  applyPatch(project: Project, patch: ProjectPatch): Project {
    return applyJsonMergePatch(project, patch)
  },
  mergePatches(patches: ProjectPatch[]): ProjectPatch {
    if (patches.length === 0) return {};
    if (patches.length === 1) return patches[0]!;
    let result: any = {};
    for (const patch of patches) {
      if (patch === null) {
        result = null;
        continue;
      }
      if (result === null) {
        continue;
      }
      if (typeof patch !== 'object' || Array.isArray(patch)) {
        result = patch;
        continue;
      }
      result = mergePatchObjects(result, patch);
    }
    return result;
  }
};
